### ISSUE TITLE
`@sanity-typed/groq`

### ISSUE THREAD

**Issue by saiichihashimoto** [saiichihashimoto/sanity-typed#52]
**Update:** 
Been working on this in the `alpha` branch! Getting it to feature complete-enough and then we'll launch this!

**Initial Description**:
With [`types`](https://github.com/saiichihashimoto/sanity-typed/tree/main/packages/types) (and previously [`schema-builder`](https://github.com/saiichihashimoto/sanity-typed/tree/main/packages/schema-builder)), typing sanity documents has become possible. However, querying for these documents always goes through a groq query, which is completely unhandled by this library. This is all due to my original use case: query for the whole document and us JS/TS to deconstruct, so using the native types was fine. For other use cases, you end up doing a lot of `Pick` and `Omit` magic along with your queries, ultimately rewriting large parts of your queries in typescript anyways.

Sanity schemas are just JS objects, so typing them was a doable challenge. The issue with typing GROQs comes from the fact that they are giant strings, similar to GraphQL queries. Deriving a type from a string is... difficult, if not impossible. Pair that with the fact that most of this library exists in an attempt to merge into sanity and we run into the few possible approaches and their major issues:

1. **Type the GROQ strings anyways.** This _might_ be possible with [typescript string literals](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) and lots of typescript magic. If possible, it would provide the best DX: a drop in replacement for `groq` or whatever is currently being used for querying. Plus, it's the closest to a native solution, which ideally means that sanity could take ownership. The issue is that the scope of this seems massive, unmaintainable, and could be impossible. Even the `types` library uses some expert typescript magic. Deriving types from a GROQ string is definitely a write-once operation, where anyone coming in to make changes will be massively overwhelmed.
2. **Codegen.** @ricokahler's work in `sanity-codegen` is the best path forward on this and it's clear that a feature complete solution is quite big (looks to be a month of dedicated work according to his estimates in https://github.com/ricokahler/sanity-codegen/issues/297). I'm generally opposed to codegen but the fact that Apollo takes this approach with GraphQL is very telling that it's likely the most viable solution. The time investment is high and the likelihood that sanity will want to maintain a codegen library is low, which ultimately means running into the same issues he did.
3. **A builder pattern library, similar to schema-builder.** This seems like the cleanest solution and the least prone to bugs. In terms of implementation, this feels like the most straightforward. A lot of the issues are similar to `schema-builder`. In `schema-builder`, rectifying sanity's schema documentation and the library's documentation was a lot of overhead for developers writing schemas and far from a drop-in solution, which left the problem feeling relatively unsolved. This is the same approach. It's definitely the farthest from a solution sanity would take ownership of, which also makes me disinterested in it. Also, I know there's a lot of work on this already (links are escaping me) so it doesn't seem worth treading the same ground.

My personal goal is to have the closest to a drop-in solution so sanity can eventually own it themselves. The issue is that these solutions each have a unique reason that sanity wouldn't (or shouldn't) maintain it internally and are the exact issues that other solutions run into. I can't think of a viable solution for this.

**Comment by saiichihashimoto** [comment]
More thoughts on this:

- No one is typing from the groq string directly. Makes sense, all the cons I mentioned are substantial.
- That package really looks like the only attempt at codegen on this. Again, makes sense.
- There's a lot more attempts at the builder pattern. Again, not my favorite, but a few packages and notable thoughts on them:
  - [sanity-typed-queries](https://github.com/danielroe/sanity-typed-queries#readme) went all in on the builder pattern. You use their builder to create the schemas, and that schema builder provides a query builder, which you use in runtime code. I like the chaining API of the query builder, although it deviates pretty far from a drop in replacement. They're including information from the schema into the builder implicitly this way, which is kind of interesting. It approaches everything from a contradictory viewpoint to this library, but there's some lessons to be learned here.
  - [groqd](https://www.npmjs.com/package/groqd) takes a whole different approach, being schema unaware and using zod in the application code to validate that it's the correct shape. This does remove the duplication of defining the same "shape" in your query and in your runtime zod and/or manual typing, but it doesn't remove the duplication of defining the same "shape" from your sanity schema. In other words, sanity-typed-queries lets you define a shape for your schema, have your queries be aware, and have your runtime objects be typed. groqd ignores the shape in the schema, so you're essentially creating the same shape in two places, having your runtime not trust your schemas. Few thoughts:
    - Again, far from a drop in replacement.
    - The idea of not trusting the sanity schema has pros and cons. The cons are obvious: now you're having to rewrite the same shape in your queries. The pro is that this is actually a safe idea. Just because your schema is a certain shape does not mean the documents in there are of that shape. You can always change your schemas drastically and the existing objects aren't invalidated, meaning you can still query for them and break your app. Deriving types from your schemas isn't always accurate, and this seems like the only approach that actually addresses that.

I'm still not too interested in the builder pattern. The idea of maintaining a separate solution (and documentation) for groqs producting correctly typed results doesn't really appeal to me. I do like that groqd does try and validate the types in runtime, and completing #81 would allow for something similar, which is interesting. That solution would have to respect however we decide to handle groqs, because we're likely to not query for documents as-is, so we'd want a zod solution that validates against the query, not the document. Again, not sure how to go about this. Part of me also thinks #82 is similar; mocks for schemas is interesting, but mocks for the actual queries used in an app allows for integration or end-to-end tests that don't have a massive amount of scaffolding. groqd has convinced me that both those issues are blocked on this one.

**Comment by saiichihashimoto** [comment]
More ramblings:

Let's just start from the ideal and work backwards into how it would get done:

```typescript
// sanity.config.ts
import { defineConfig, defineType, ... } from "@sanity-typed/types";
import type { InferSchemaValues } from "@sanity-typed/types";

const config = defineConfig(...);

export default config;
export type SanityTypes = InferSchemaValues<typeof config>;
```

```typescript
import { groq } from "@sanity-typed/groq";
import { sanityZod } from "@sanity-typed/zod";
import config, { SanityTypes } from "???/sanity.config.ts";

// This would be a branded string carrying a type? Something like that?
// Whatever it is, it wouldn't result in carrying SanityTypes, but the type corresponding to the actual query.
const query = groq<SanityTypes>(config)`...`;

// It needs the runtime config and query to generate the zod that fits the query results.
// We can't use SanityTypes, because typescript can't actually affect runtime like that
const zodSchema = sanityZod(config, query);

// docs is untyped...
const docs = await client.fetch(query);
// ...validatedDocs is correctly typed and runtime validated
const validatedDocs = zodSchema.parse(docs);

// It's unfortunate that zod is required. Not the goal of this library to force that, although it is substantially safer.
// Somehow, something like this should work...
const castedTypesDocs = await client.fetch(query) as SomeMagicHere<typeof query>;

// I'd love to avoid wrapping the client library as well, but if we did, then it could "just work"
// client.fetch would have a signature like client.fetch = <Generic>(query: QueryOf<Generic>) => Generic;
// This could be done with augmentation too, which is actually nice
const assumedTypeDocs = await client.fetch(query);
// This is much closer to the actual solution I want.
```

The real answer is that I can't have it all.

There isn't a solution that will be a drop in replacement into sanity that handles groqs, the document queries, and zods so cleanly. Ultimately, if I accepted that I have to have some kind of builder pattern for groqs, my life becomes substantially easier for everything else, but it ends up just being a sanity-typed-queries or groqd replacement. I'm not interested in trying another solution for the same approach, they seem like the best way forward for that.

I think the reality is that codegen and the builder pattern already have the best versions present, codegen being on hiatus because its a very difficult solution and the builder pattern having active communities on it. If there was a way to somehow work in a provided type to the builder pattern (ie the type generated from `@sanity-typed/types` into groqd), then we could close the loop on that. The only real way this library moves forward in a way that isn't treading the same ground is to take a stab at the typescript string parsing solution for groqs, and to use a runtime regex on the the groqs to produce zods, and to keep the two types in sync. This feels like the land of a million bugs, but honestly the only approach that isn't just trying to do groqd but "better."

Due to these issues, I'm not going to prioritize this. It seems like a massive initiative that ultimately might not pay off, so I'm going to push on other things first.

**Comment by saiichihashimoto** [comment]
Some links:

- https://github.com/sanity-io/groq-js Seems like a js runtime solution for this, whereas we want a ts only solution. Might be good for some kind of merged solution, since using it might get us 95% of what we want from zod.
- https://sanity-io.github.io/GROQ/ If we write the typescript string parsing, the spec should be very useful for creating the finite automata
- https://github.com/sanity-io/groqfmt Seems irrelevant, but kind of fun

**Comment by SamScott3000** [comment]
Played around with trying to type query strings. A query formatter could be useful, but is less drag and drop.
Any more thoughts on which route to go down?

```ts
type QueryStringLiteralToType<
  Q extends string,
  T extends Record<string, any>,
> = Q extends `*[_type == '${infer QueryType}']${infer NumericString}`
  ? QueryType extends T["_type"]
    ? NumericString extends `[${number}]`
      ? T
      : T[]
    : never
  : never;

// Usage examples (you would probably want to select from all types based on the _type value):
const type1: QueryStringLiteralToType<"*[_type == 'footer']", FooterType> = []; // type: FooterType[]
const type2: QueryStringLiteralToType<"*[_type == 'footer'][0]", FooterType> = {}; // type: FooterType
const type3: QueryStringLiteralToType<"*[_type == 'WRONG SLUG'][0]", FooterType> = {}; // type: never
```

**Comment by saiichihashimoto** [comment]
This is starting to look good! I think, if we were to go the string route, we'd want it to be as one-to-one as the current API. So I'm thinking about how we'd start with this:

```typescript
import { createClient } from '@sanity/client';

const client = createClient({
  // ...
});

// Somehow, include types derived from schema into this.
const docs = await client.fetch('SOME GROQ QUERY');
```

Few solutions. The easiest is to create our own export and use it:

```typescript
import { createClient } from '@sanity-typed/client';
import type { Values } from '../../../../../../../wherever/your/schema/is';

// We COULD include `Values` into `createClient`, so all client methods can be type aware. Might be better that way. eg:
// const client = createClient<Values>({
const client = createClient({
  // ...
});

// client.fetch would use the string magic to know the correct types for `docs`
const docs = await client.fetch<Values>('SOME GROQ QUERY');
```

We could also find a way to augment `sanity`'s own `createClient` to do what we wanted. I think augmenting, is easier to migrate at the end of the day, but can run into a lot of its own issues, and might become incompatible over time.

Back to your example @SamScott3000.

1. I think you have the right idea.
2. Internally, we'd have a generic like you're doing that receives a query and the available Values, and would operate on those.
3. I think we'd probably want to split this into `@sanity-typed/client` and `@sanity-typed/groq` since we'd likely end up doing more magic on the client other than groqs, and vice versa with the groqs. Implementation details, I'm just mentioning them so I remember them myself.
4. I'd say the ideal is to not pass it `FooterType` specifically, but pass it `Values`. A query can operate on all available types, so ideally we're doing this to everything possible and then narrowing down.
5. The biggest worry I have is that, to handle all use cases, we'd likely want to look at the [GROQ FMA](https://sanity-io.github.io/GROQ/) and recreate that in typescript. Unless we do this, we're always only partially recreating the language. Sounds like a lot of work lol

**Comment by saiichihashimoto** [comment]
If anyone's watching this, the alpha branch is my first stab at this. I'm still missing a lot but... it's generally working.
